src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py

```diff
@@ -62,15 +65,38 @@ def generate(
         component_graph = connection_manager.get().component_graph
 
         battery_inverters = list(
-            next(iter(component_graph.predecessors(bat_id))) for bat_id in component_ids
+            frozenset(
+                filter(
+                    component_graph.is_battery_inverter,
+                    component_graph.predecessors(bat_id),
+                )
+            )
+            for bat_id in component_ids
         )
 
-        if len(component_ids) != len(battery_inverters):
+        if not all(battery_inverters):
             raise ComponentNotFound(
-                "Can't find inverters for all batteries from the component graph."
+                "All batteries must have at least one inverter as a predecessor."
+            )
+
+        all_connected_batteries = set()
+        for inverters in battery_inverters:
+            for inverter in inverters:
+                all_connected_batteries.update(
+                    component_graph.successors(inverter.component_id)
+                )
+
+        if len(all_connected_batteries) != len(component_ids):
+            raise FormulaGenerationError(
+                "All batteries behind a set of inverters must be requested."
             )
 
-        for idx, comp in enumerate(battery_inverters):
+        builder.push_oper("(")
+        builder.push_oper("(")
+        # Iterate over the flattened list of inverters
+        for idx, comp in enumerate(
+            inverter for inverters in battery_inverters for inverter in inverters
+        ):
```

Comment:
maybe the inverter ids can be collected into a set before enumerating, to deduplicate?

```suggestion
        for idx, comp in enumerate(
            set(inverter for inverters in battery_inverters for inverter in inverters)
        ):

Reply:
I made `battery_inverters` directly to a `frozenset` instead